feat(quotas): Separate processing and indexing quotas

jan-auer opened the pull request:
Once metrics have been extracted, transaction events need to have the
new `DataCategory::TransactionIndexed` quota available to ingest and
store the payload.

The new indexing quota replaces the previous `DataCategory::Transaction`
for transaction payloads. Transaction quotas now cover all ingested
transactions that have metrics extracted. Transaction quota however is
only counted in the metrics part and not in event ingestion.

# Background

Rate limiting and quota enforcement has to apply to _legacy_ plans that
did not have processed transaction functionality, and to _new_ plans
that introduce them.

### Organization on Legacy Plan

There is a single quota for total transactions. We need to extract
metrics for indexed transactions so that there is backfilled history at
the time the customer switches to the new plan. This requires to extract
metrics for all accepted transactions.

### Organization on New Plan

This plan has a dedicated additional quota for indexed transactions.

- Validation and inbound data filters do not consume any quota.
  There are no metrics extracted for such dropped events.
- If there is transaction quota available, extract and ingest metrics for
  this transaction. Otherwise, drop the transaction event and any
  transaction metrics bucket.
- Dynamic sampling may drop (previously processed) transaction events.
  Extracted metrics and the transaction quotas are not
  impacted by this.
- If there is additional indexing quota available, store the
  sampled transaction event payload. Otherwise, drop the transaction event.

# Algorithm

**Flow:**

1. Check if metrics transaction is enabled. 
   1. If disabled, rate limit the transaction item with the `transaction` data category and exit.
   2. Otherwise, continue with step 2.
2. Check for transaction quota but do not increment it.
   1. If quota is exceeded, drop the item and all metrics. 
   2. Otherwise, continue with step 3.
3. Check if metrics have been extracted for this transaction yet.
   1. If not, skip to step X
   2. Otherwise, keep the item. Metrics can be extracted now.
4. Metrics have been extracted. Check for indexing quota.
   1. If quota is exceeded, drop the item but keep metrics.
   2. Otherwise, keep the item. The transaction should be stored.

Transaction quota is incremented by the extracted metrics.

# References

- Requires https://github.com/getsentry/relay/pull/1535
- Requires https://github.com/getsentry/getsentry/pull/8642
- Replaces https://github.com/getsentry/relay/pull/1517

jjbayer committed:
feat: DataCategory::TransactionIndexed

jjbayer committed:
fix: doc syntax

jan-auer committed:
feat(quotas): Separate processing and indexing quotas

jan-auer committed:
fix(quotas): Do it correctly

jan-auer committed:
meta: Changelog

olksdr commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -246,9 +255,23 @@ pub struct Enforcement {
 }
 
 impl Enforcement {
+    /// Returns true if the event should be rate limited.
```
```suggestion
    /// Returns `true` if the event should be rate limited.
```

to keep it consistent

jjbayer commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -152,6 +157,10 @@ impl EnvelopeSummary {
                 summary.has_plain_attachments = true;
             }
 
+            if *item.ty() == ItemType::Transaction && item.metrics_extracted() {
```
nit: As the field is now called `event_metrics_extracted`, should we allow item type `Event` here as well?

jan-auer replied:
We would probably also have to do this for `Security` and all other event-like categories. I'm not sure exactly how we'd want those to behave yet, which is why I would stick with just `Transaction` for the detection.

jjbayer commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -246,9 +255,23 @@ pub struct Enforcement {
+    /// Returns `true` if metrics extracted from the event should be rate limited.
+    #[cfg(feature = "processing")]
+    pub fn event_metrics_active(&self) -> bool {
+        // TODO: Rename this
```
```suggestion
```

jjbayer commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -378,9 +425,29 @@ where
+                // quota. Quota will be consumed by metrics in the metrics aggregator instead.
+                event_limits = (self.check)(scoping.item(category), 0)?;
+                longest = event_limits.longest();
+                enforcement.event_metrics = CategoryLimit::new(category, 1, longest);
```
This `1` is only here such that `is_active()` evaluates to `true`, right? It might as well be `2` or `999`.

iker-barriocanal replied:
IIRC, no; the `1` is because there's only one transaction. There's no case in which we use any other number than `1`, but `0` in this PR. You could argue that there's no point in having the API designed that way, and that's a valid point.

jan-auer replied:
This is basically it. The original idea was to create a non-optional type that we assign unconditionally, and that carries all information about the enforced rate limit. The only exposed method ended being `is_active`, which checks if we have a non-zero quantity. It's factually correct here, since we are in fact limiting `1` transaction.

iker-barriocanal commented on relay-cabi/include/relay.h:
```diff
@@ -49,9 +49,17 @@ enum RelayDataCategory {
+  /**
+   * Indexed transaction events.
+   * This is the category for transaction payloads that were stored, that is, they were not
+   * filtered out by Dynamic Sampling.
```
```suggestion
   * Indexed transaction events.
   * This is the category for transaction payloads that were stored, that is, they were not
   * filtered out by filters such as Dynamic Sampling.
```
Dynamic sampling is not the only mechanism to avoid storing transactions.

jan-auer replied:
Let me actually rewrite this. Dynamic sampling isn't really a filter -- and inbound filters would affect both data categories alike. The new text is going to be:

```
    /// Indexed transaction events.
    ///
    /// This is the category for transaction payloads that were accepted and stored in full. In
    /// contrast, `transaction` only guarantees that metrics have been accepted for the transaction.
```

iker-barriocanal commented on relay-common/src/constants.rs:
```diff
@@ -113,8 +113,14 @@ pub enum DataCategory {
     TransactionProcessed = 8,
+    /// Indexed transaction events.
+    /// This is the category for transaction payloads that were stored, that is, they were not
+    /// filtered out by Dynamic Sampling.
```
Same suggestion as the previous.

iker-barriocanal commented on relay-server/src/utils/metrics_rate_limits.rs:
```diff
@@ -51,13 +51,13 @@ impl BucketLimiter {
                 if mri.name == "duration" {
-                    // The "duration" metric is extracted exactly once for every processed transaction,
-                    // so we can use it to count the number of transactions.
+                    // The "duration" metric is extracted exactly once for every processed
+                    // transaction, so we can use it to count the number of transactions.
```
Since this PR also implements the renaming dance of transactions, indexed transactions, and processed transactions:
```suggestion
                    // The "duration" metric is extracted exactly once for every
                    // transaction, so we can use it to count the number of transactions.
```

iker-barriocanal commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -246,9 +255,23 @@ pub struct Enforcement {
+        self.event.is_active()
+    }
+
+    /// Returns `true` if metrics extracted from the event should be rate limited.
+    #[cfg(feature = "processing")]
+    pub fn event_metrics_active(&self) -> bool {
```
The reason to enable the processing flag here is for optimization reasons, right? I see it's only used in `enforce_quotas`.

jan-auer replied:
The compiler would also complain that this function is dead code.

iker-barriocanal replied:
Hmmm, smart compiler.

iker-barriocanal commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -378,9 +425,29 @@ where
+
+                // If the main category is rate limited, we drop both the event and metrics. If
+                // there's no rate limit, check for specific indexing quota and drop just the event.
+                if summary.event_metrics_extracted && longest.is_none() {
```
I want to double-check if I understand this correctly. The first time we get here is from the "fast path", and metrics aren't extracted at that point; which means we don't care about indexing transactions for now. Then, we process the envelope, which includes the metric extraction. The second time we get here, in `enforce_quotas` iirc, the envelope have almost been fully processed but the transaction isn't indexed yet, so we do care about extracting metrics and we'd get inside this `if` branch.

Is this correct?

iker-barriocanal commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -850,4 +936,110 @@ mod tests {
+        let (enforcement, limits) = limiter.enforce(&mut envelope, &scoping()).unwrap();
+
+        // NOTE: Since metrics have not been extracted on this item, we do not check the indexing
+        // quota. Basic processing quota is not denied, so the item must pass rate limiting.
```
Is it worth noting that the transaction is rate limited at a later point? Your call.

iker-barriocanal commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -850,4 +936,110 @@ mod tests {
+    }
+
+    #[test]
+    fn test_enforce_transaction_attachment_enforced_metrics_extracted() {
```
```suggestion
    fn test_enforce_transaction_attachment_enforced_metrics_extracted_indexing_quota() {
```
not sure? To be consistent with other test names.

iker-barriocanal commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -850,4 +936,110 @@ mod tests {
+        let mut mock = MockLimiter::default().deny(DataCategory::Transaction);
+        let limiter = EnvelopeLimiter::new(Some(&config), |s, q| mock.check(s, q));
+
+        let (enforcement, _limits) = limiter.enforce(&mut envelope, &scoping()).unwrap();
```
Why don't we check the limits in this test and in the one below?

jan-auer replied:
This test is mostly about asserting correct calls to the inner rate limiter and there are already tests that then check the response behavior. We can keep this test to a single concern with fewer assertions.

iker-barriocanal commented on tests/integration/test_store.py:
```diff
@@ -649,10 +649,70 @@ def send_buckets(buckets):
+    key_id = mini_sentry.get_dsn_public_key_configs(project_id)[0]["numericId"]
+    projectconfig["config"]["quotas"] = [
+        {
+            "id": "test_rate_limiting_{}".format(uuid.uuid4().hex),
```
nit: f-strings
```suggestion
            "id": f"test_rate_limiting_{uuid.uuid4().hex}",
```

iker-barriocanal commented on tests/integration/test_store.py:
```diff
@@ -649,10 +649,70 @@ def send_buckets(buckets):
+            "reasonCode": "get_lost",
+        },
+        {
+            "id": "test_rate_limiting_{}".format(uuid.uuid4().hex),
```
nit: f-strings
```suggestion
            "id": f"test_rate_limiting_{uuid.uuid4().hex},
```

jan-auer committed:
ref: Address review comments

olksdr reviewed (approved):
still lgtm

jjbayer committed:
feat(quotas): Rate limit metrics before aggregator (#1540)

### Background

In https://github.com/getsentry/relay/pull/1515 we implemented rate
limiting functionality for metrics buckets, which are applied after
flushing them from the metrics aggregator by checking redis.

### This PR

When a quota is already exhausted, and that information has already been
cached on the project state, there is no need to go through the
aggregator at all. Instead, we apply the rate limiting logic _before_
sending metrics or buckets into the aggregator.

### Implementation details

The utility type `BucketLimiter` was converted to a generic
`MetricsLimiter<T>`, where `T` can be either `Metric` or `Bucket`.

### Notes

Will be merged into https://github.com/getsentry/relay/pull/1537 &
deployed together.

jjbayer committed:
Merge branch 'master' into feat/indexed-transactions-limiting

jan-auer committed:
Merge branch 'master' into feat/indexed-transactions-limiting

* master:
  fix(gha): Skip Sentry integration tests on library release builds (#1550)
  ref: Use wildcard re-exports where applicable (#1526)
  ci(gha): Skip Sentry integration tests on library release builds (#1549)
  release: 0.8.15
  fix(py): Respect the renormalize flag (#1548)
  (fix)e2e: Use report self hosted issues env variable (#1539)
  meta(vscode): Enable all features in Rust-Analyzer (#1542)
  release: 0.8.14
  build(craft): Fix manylinux artifact name (#1547)
